jwt/src/main/java/me/uport/sdk/jwt/JWTTools.kt

```diff
@@ -190,59 +192,73 @@ class JWTTools(
-            val matches = ddo.publicKey.filter {
-                it.type != DelegateType.Curve25519EncryptionPublicKey
-            }.map { pubKeyEntry ->
+        val matches = publicKeys.map { pubKeyEntry ->
 
-                val pkBytes = pubKeyEntry.publicKeyHex?.hexToByteArray()
-                        ?: pubKeyEntry.publicKeyBase64?.decodeBase64()
-                        ?: pubKeyEntry.publicKeyBase58?.decodeBase58()
-                        ?: ByteArray(PUBLIC_KEY_SIZE)
-                val pubKey = PublicKey(pkBytes.toBigInteger()).normalize()
+            val pkBytes = pubKeyEntry.publicKeyHex?.hexToByteArray()
+                    ?: pubKeyEntry.publicKeyBase64?.decodeBase64()
+                    ?: pubKeyEntry.publicKeyBase58?.decodeBase58()
+                    ?: ByteArray(PUBLIC_KEY_SIZE)
+            PublicKey(pkBytes.toBigInteger()).normalize()
 
-                (pubKeyEntry.ethereumAddress?.clean0xPrefix() ?: pubKey.toAddress().cleanHex)
+        }.filter { publicKey ->
 
-            }.filter { ethereumAddress ->
+            ecVerify(messageHash, sigData, publicKey)
+        }
+
+        return matches.isNotEmpty()
+    }
 
-                //this method of validation only works for uPort style JWTs, where the publicKeys
-                // can be converted to ethereum addresses
-                ethereumAddress.toLowerCase() == recoveredAddress
-            }
+    private fun verifyRecoverableES256K(publicKeys: List<PublicKeyEntry>, sigData: SignatureData, signingInputBytes: ByteArray): Boolean {
 
-            if (matches.isNotEmpty()) {
-                return payload
-            }
+        val recoveredPubKey: BigInteger = try {
+            signedJwtToKey(signingInputBytes, sigData)
+        } catch (e: Exception) {
+            BigInteger.ZERO
         }
 
-        throw InvalidJWTException("DID document for ${payload.iss} does not have any matching public keys")
+        val pubKeyNoPrefix = PublicKey(recoveredPubKey).normalize()
+        val recoveredAddress = pubKeyNoPrefix.toAddress().cleanHex.toLowerCase()
+
+        //TODO: this check needs to be adapted to the logic from [did-jwt](https://github.com/uport-project/did-jwt/blob/3ea977934e844598b2bc6576369335fd1972a12a/src/JWT.js#L118)
+        val matches = publicKeys.map { pubKeyEntry ->
+
+            val pkBytes = pubKeyEntry.publicKeyHex?.hexToByteArray()
+                    ?: pubKeyEntry.publicKeyBase64?.decodeBase64()
+                    ?: pubKeyEntry.publicKeyBase58?.decodeBase58()
+                    ?: ByteArray(PUBLIC_KEY_SIZE)
+            val pubKey = PublicKey(pkBytes.toBigInteger()).normalize()
+
+            (pubKeyEntry.ethereumAddress?.clean0xPrefix() ?: pubKey.toAddress().cleanHex)
+
+        }.filter { ethereumAddress ->
+
+            //this method of validation only works for uPort style JWTs, where the publicKeys
```

Comment:
This comment is no longer relevant since this method is only being called for uPort style JWT based on header info